### DISCUSSION TITLE
Rewriting for a URL structure

### DISCUSSION THREAD

**Discussion by rinkusam12** [vercel/next.js#22219] [Help]
I don't know if this can be considered a bug or not. Basically, I am using the rewrite feature of NextJS to rewrite /category/[slug] to /[slug].
I have following route setup in NextJS project:
index.js //Homepage
[product].js // for the product detail page
/category/[main_category].js // For main category page

Next Config JS consist of rewrite config as follow:
module.exports = {
  async rewrites() {
    return [
      {
        source: "/:main",
        destination: "/category/:main",
      },
    ];
  },
};

Now the problem I am facing is in development mode it doesn't work which is okay. But when I create the next build and run it on production mode in local, I am able to open the category main page on the desired URL but the product detail page is not accessible at all.

For some reason, it works when I deploy it to vercel, but there also if I add up one more rewrite suppose /local/[location] to /[location], then it doesn't work for the /local routes. Only the /category/:main are getting rewrites(due to the rule being on top) and others are getting skipped.

**Comment by mkrtchian** [comment]
Hello,

You say you would like to "to rewrite /category/[slug] to /[slug]", and you provide a code that rewrites `/:main` to `/category/:main`. I guess one of these was an error, but anyway there is another problem in your approach I think.

When I try locally, even in dev mode:
* if `/:main` is rewritten to `/category/:main`, I get the content of /category/[slug].js on `/category/a`, `/category/b` but also `/a` or `/b`, and I never get the content of /[product].js
* if `/category/:main` is rewritten to `/:main`, I get the content of /[product].js  on `/category/a`, `/category/b`, but also `/a`, or `/b`, and I never get the content of /category/[main_category].js

The problem here is that when you provide a file like [slug].js, or when you rewrite /:main to somewhere else, it means that you target "all the possible slugs that are between two / signs".
If all the possibilities of your root URL (`/a`, `/b`, `/c` etc.) refer to categories, they can't refer to products at the same time, and the same in the other direction.

**Reply by rinkusam12** [reply]
@mkrtchian 
Ok, I might have confused with the setup part but basically what I want to achieve goes something like this:

Let say I have the category "flower". In my NextJS project file that serves category page is /category/[slug].js. So it ends up with a /category/flower route.

Now let say I have a product "red-wedding-flower". For that, I have /[product].js in the root of the page directory. So it ends up with a route of "/red-wedding-flower".

Now using the rewrite functionality, I want to rewrite the route of /category/flower to /flower. Basically, I want to omit the category from the URL.

Now the question might come that, what if slug flower also comes as a product. For that, I have implemented the backend to have a unique slug for the category and product.

Also, the rewrite is refrence from the NextJS doc at [https://nextjs.org/docs/api-reference/next.config.js/rewrites#incremental-adoption-of-nextjs](https://nextjs.org/docs/api-reference/next.config.js/rewrites#incremental-adoption-of-nextjs)
In the doc, 
async rewrites() {
    return [
      // we need to define a no-op rewrite to trigger checking
      // all pages/static files before we attempt proxying
      {
        source: '/:path*',
        destination: '/:path*',
      },
      {
        source: '/:path*',
        destination: `/category/:path*`,
      },
    ]
  },

As per my understanding, here we are proxying all the /:path* to /category/:path*, if we don't find a matching at the root destination. 

I hope I made myself clear. This is something related to SEO. In Wordpress we used to achieve this using htaccess rewrites.

PS: 
I am using the ISG method to statically generate the page at build time.

**Reply by mkrtchian** [reply]
Unfortunately to my understanding the next.js router has no way to know if a particular slug exists as a category or not, because you have to ask your backend for that, and to ask your backend you have to enter in the code of your page.

You say "here we are proxying all the /:path* to /category/:path*, if we don't find a matching at the root destination", but we do always find a match, because [product].js matches everything at root level, not only existing product slugs.

A solution may be having a unique page handling everything at root level ([product].js that you may rename for semantic reasons). In that file you can do the fetch from your backend, and depending on the result, display the category component or the product component. And in your /category/[slug].js page, you can just display the category component.
